### PR TITLE
Combine all the array benchmarks to reduce duplication 

### PR DESCRIPTION
**Pull request by adithyaov**
Use --stream-size for all the array benchmarks now

Replace the following files:
- benchmark/Streamly/Benchmark/Data/Array.hs
- benchmark/Streamly/Benchmark/Data/ArrayOps.hs
- benchmark/Streamly/Benchmark/Data/Prim/Array.hs
- benchmark/Streamly/Benchmark/Data/Prim/ArrayOps.hs
- benchmark/Streamly/Benchmark/Data/SmallArray.hs
- benchmark/Streamly/Benchmark/Data/SmallArrayOps.hs
- benchmark/Streamly/Benchmark/Memory/Array.hs
- benchmark/Streamly/Benchmark/Memory/ArrayOps.hs

With:
- benchmark/Streamly/Benchmark/Array.hs
- benchmark/Streamly/Benchmark/ArrayOps.hs

### CODE REVIEW

**Comment by adithyaov** [comment]
> Are the benchmark results before and after this change identical?

Things seem to have improved. No difference in allocation so it could just be a measurement error.
```
x = after change             y - x     x - x
--------------------------- -------  ----------
elimination/read            +129.77  -Infinity
transformation/scanl'       +129.60     +28.12
generation/writeN . unfoldr +107.63  -Infinity
elimination//=               +89.83     +34.13
transformationX4/map         +65.79     +20.00
elimination/==               +53.40     -17.09
elimination/toStreamRev      +46.71     -28.98
```

File: benchmark/Streamly/Benchmark/Data/Array.hs [hunk 1]
```diff
@@ -4,99 +4,191 @@
...
 -------------------------------------------------------------------------------
 --
 -------------------------------------------------------------------------------
 
+#ifdef MEMORY_ARRAY
+-- Drain a source that generates a pure array
+{-# INLINE benchPureSrc #-}
+benchPureSrc ::
+       NFData a =>
+#ifdef MEMORY_ARRAY
```
Comment (harendra-kumar) [inline benchmark/Streamly/Benchmark/Data/Array.hs:44]: We are already inside `#ifdef MEMORY_ARRAY`.

File: benchmark/Streamly/Benchmark/Data/Array.hs [hunk 2]
```diff
@@ -4,99 +4,191 @@
...
 
-import Streamly.Benchmark.Common
+import Streamly.Benchmark.Common hiding (benchPureSrc)
 import Gauge
 
+#ifdef MEMORY_ARRAY
+import Foreign.Storable (Storable(..))
+#endif
+
+#ifdef DATA_PRIM_ARRAY
```
Comment (harendra-kumar) [inline benchmark/Streamly/Benchmark/Data/Array.hs:31]: You could combine all imports for DATA_PRIM_ARRAY in a single conditional, and similarly for MEMORY_ARRAY.

File: benchmark/Streamly/Benchmark/Data/Array.hs [hunk 3]
```diff
@@ -4,99 +4,191 @@
...
-        -- , benchPureSink "length . IsList.toList" (length . GHC.toList)
-        , benchIOSink "foldl'" Ops.pureFoldl'
-        , benchIOSink "read" (S.drain . S.unfold A.read)
-        , benchIOSink "toStreamRev" (S.drain . A.toStreamRev)
+        , benchPureSink value "length . IsList.toList" (length . GHC.toList)
+#endif
+        , benchIOSink value "foldl'" Ops.pureFoldl'
+        , benchIOSink value "read" Ops.unfoldReadDrain
+        , benchIOSink value "toStreamRev" Ops.toStreamRevDrain
+#ifndef DATA_PRIM_ARRAY
```
Comment (harendra-kumar) [inline benchmark/Streamly/Benchmark/Data/Array.hs:134]: You could use `#if !defined(DATA_PRIM_ARRAY) && defined(DEVBUILD)`.

File: benchmark/Streamly/Benchmark/Data/ArrayOps.hs [hunk 4]
```diff
@@ -37,45 +60,83 @@ value = 100000
...
 -------------------------------------------------------------------------------
 
-type Stream = A.Array
+-------------------------------------------------------------------------------
+-- CPP Common to:
+-- MEMORY_ARRAY
+-- DATA_ARRAY
+-- DATA_PRIM_ARRAY
+-- DATA_SMALLARRAY
+-------------------------------------------------------------------------------
```
Comment (harendra-kumar) [inline benchmark/Streamly/Benchmark/Data/ArrayOps.hs:69]: These comments and other such comments can perhaps be removed, they don't seem to be useful.

File: benchmark/Streamly/Benchmark/Data/Array.hs (hunk 1 above)
Comment (adithyaov) [reply benchmark/Streamly/Benchmark/Data/Array.hs:44]: Oh Yeah. My bad.

File: benchmark/Streamly/Benchmark/Data/Array.hs (hunk 3 above)
Comment (adithyaov) [reply benchmark/Streamly/Benchmark/Data/Array.hs:134]: I didn't rewrite on purpose. I'm thinking of the constants that define what type of array this module is aimed for as top-level conditionals

File: benchmark/Streamly/Benchmark/Data/ArrayOps.hs (hunk 4 above)
Comment (adithyaov) [reply benchmark/Streamly/Benchmark/Data/ArrayOps.hs:69]: Once the module grows, CPP gets really messy. As long as these comments aren't negatively affecting the code, I am not in favor of removing them. They are comments for the developers anyway.

**Pull request merged**